### PR TITLE
Beam radius in ILL SANS reduction

### PR DESCRIPTION
**Pull request by gui-co**
**Description of work.**

This PR adds the possibility to set a different beam radius for each measured distance in ILL SANS data reduction.

One can now set the `BeamRadius` property to a single value or an array of value (same length as `SampleRuns`). A new property has been added to set the beam radius used during transmission experiments: `TransmissionBeamRadius`

**To test:**

With data from `ExternalData/Testing/Data/SystemTest/ILL/D11`

```python
SANSILLAutoProcess(
    SampleRuns="2889,2885,2881",
    SampleTransmissionRuns="2871",
    BeamRadius="0.1,0.1,0.1",
    TransmissionBeamRadius=0.1,
    TransmissionBeamRuns="2867+2868",
    OutputWorkspace="ws"
)
```

A warning is raised if the number of beam radius is not 1 and different from the number of SampleRuns.

Part of #26906 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DannyHindson** [comment]
Just started having a look at this. I get the following error when I run your test script. Am I doing something wrong? Looks like this is from one of the calls to the child SANSILLReduction algorithm:

Invalid value for BeamInputWorkspace: Beam input workspace is mandatory for transmission calculation.
Error in execution of algorithm SANSILLAutoProcess:
Some invalid Properties found
  at line 371 in '/home/danny/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py'
  caused by line 502 in '/home/danny/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py'
  caused by line 1124 in '/home/danny/mantid/Framework/PythonInterface/mantid/simpleapi.py'
  caused by line 795 in '/home/danny/mantid/Framework/PythonInterface/mantid/simpleapi.py'
RuntimeError: Some invalid Properties found
  at line 371 in '/home/danny/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py'
  caused by line 502 in '/home/danny/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py'
  caused by line 1124 in '/home/danny/mantid/Framework/PythonInterface/mantid/simpleapi.py'
  caused by line 795 in '/home/danny/mantid/Framework/PythonInterface/mantid/simpleapi.py'
  File "<string>", line 11, in <module>
  File "/home/danny/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 1120, in __call__
    algm.execute()

**Comment by gui-co** [comment]
Thanks for the comment. If forgot a parameter, sorry for that. 
I updated the comment with a new test script:
```python
SANSILLAutoProcess(
    SampleRuns="2889,2885,2881",
    SampleTransmissionRuns="2871",
    BeamRadius="0.1,0.1,0.1",
    TransmissionBeamRadius=0.1,
    TransmissionBeamRuns="2867+2868",
    OutputWorkspace="ws"
)
```

File: docs/source/release/v6.0.0/sans.rst
Comment (DannyHindson) [inline docs/source/release/v6.0.0/sans.rst:18]: Please can you also mention that you've added a new TransmissionBeamRadius parameter + the default value for BeamRadius has changed from 0.05 to 0.1.

Also please could you make one other documentation update. I noticed there's a screenshot in the file drill-settings.png which shows the settings screen on the DrILL user interface. It would be useful to update because this screen contains the new TransmissionBeamRadius setting

**Review by DannyHindson — CHANGES_REQUESTED** [review]
Thanks for the updated test script. That works fine now and I see the validation working when the beam radius list has more than one value.

Please can you make the couple of documentation changes I've described. Also I don't see a unit test for the SANSILLAutoProcess algorithm in the Framework\PythonInterface\test\python\plugins\algorithms\WorkflowAlgorithms folder. Please could you create one and add a test for the validation on the beam radius property? You could use self.assertRaises(ValueError,....). It's perhaps a bit much to ask you to fill out the whole unit test suite based on this relatively small change but creating a test just for this one feature will mean future changes also get a unit test added

File: docs/source/release/v6.0.0/sans.rst
Comment (gui-co) [reply docs/source/release/v6.0.0/sans.rst:18]: I updated the release note to mention the new parameter and the default value. I also changed the settings window screenshot.

**Comment by gui-co** [comment]
Thanks for the comments. The documentation has been updated. I also added a unit test for SANSILLAutoProcess which tests the input validation.

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py
Comment (DannyHindson) [inline Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py:26]: Please can you correct the typo - I think this should be test_validateInput

**Review by DannyHindson — CHANGES_REQUESTED** [review]
Just one v minor thing. Please can you add a 'p' to the new unit test name

Also I noticed something that I don't is related to your work so will raise a separate issue. If I open up workbench, run your test script and then make any edit to the script (eg edit the beam radius parameter or even just add a carriage return somewhere), I get this error reported in the log:

image

This message looks like it's come from the MaxEnt algorithm's init() method. It seems like something in workbench script editor is running the init method on this algorithm (perhaps it does it on all the algorithms) and if there's a workspace present with uneven bins then you get this error (MaxEnt doesn't like workspaces with uneven bins). I think the only connection with your work is that the SANSILLAutoProcess algorithm outputs a workspace with uneven bins that ends up being the default value for one of MaxEnt's input parameters.

**Review by DannyHindson — CHANGES_REQUESTED** [review]
Just one v minor thing. Please can you add a 'p' to the new unit test name

Also I noticed something that I don't is related to your work so will raise a separate issue. If I open up workbench, run your test script and then make any edit to the script (eg edit the beam radius parameter or even just add a carriage return somewhere), I get this error reported in the log:

image

This message looks like it's come from the MaxEnt algorithm's init() method. It seems like something in workbench script editor is running the init method on this algorithm (perhaps it does it on all the algorithms) and if there's a workspace present with uneven bins then you get this error. I think the only connection with your work is that the SANSILLAutoProcess algorithm outputs a workspace with uneven bins. I was running on Ubuntu when I saw this

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py
Comment (gui-co) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py]: Done.

**Comment by gui-co** [comment]
I have already seen the error you mention. But I am not not able to reproduce it now...
(Ubuntu 18.04 as well)

**Review by DannyHindson — DISMISSED** [review]
Thanks for making the changes, all looks good now

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py
Comment (gvardany) [inline Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py:12]: I'm not sure if this is needed, validateInputs we can trust, but if we don't then we have to cover all possible permutations which is not feasible.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:126]: Are you allowing for more than one transmission in this PR? (i.e. different wavelengths). If not, can you explain what constraints are you imposing on the transmission runs?

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py
Comment (gvardany) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py:12]: The actual execution takes some time, so is covered in the system tests.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py
Comment (gui-co) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:126]: No, it is still single transmission.
I did not change anything about the constraints on the transmission runs. There should be one transmission only.

Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:126]: So if there are many with **,** it will raise an error instead of replacing **,** with **+** ?

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py
Comment (DannyHindson) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py:12]: Two more thoughts from me on this:
- My default approach on any new algorithm is to implement a unit test. I like the clarity of this because it avoids a grey area about what code we trust\don't trust - and with lots of people joining\leaving the team I think the clarity is helpful
- I think there are lots of unit tests which don't provide 100% coverage of all possible inputs. My view was that partial coverage was better than no coverage though

@gvardany - since you're the gatekeeper on this you probably have the final say but thought I'd come back with those comments at least

Comment (gvardany) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessTest.py:12]: I agree in general. We should add a unit test that ensures the minimum options run is operational, @gui-co can you add one? But trying to cover validateInputs is simply impractical for this type of master algorithms.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py
Comment (gui-co) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:126]: Yes, a warning message to invite the user to use `+` instead of `,`

**Review by DannyHindson — COMMENTED** [review]
Not sure if this PR needs to come back to me for approval but I'm a bit lost.
It seems like we used to have a unit test that completely covered the following validation on ~10 of the input parameters:
 "Wrong number of {0} runs: {1}. Provide one or as many as sample runs"
...but now we've deleted it. Isn't that a backwards step? I don't understand why it's better to trust that the code on line 143-169 of SANSILLAutoProcess.py works. This validation code only exists in this workflow algorithm not in any of the child algorithms

**Comment by gvardany** [comment]
We will add another test suite exclusively for validateInputs. I thought it can be done in a separate PR, but we could do here at the end.

**Comment by gui-co** [comment]
For SANSILLAutoProcess, there is now:
* a unit test that checks the processing
* a unit test that checks the inputs

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessInputTest.py
Comment (gvardany) [inline Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcessInputTest.py:26]: Can you split this into individual tests with representative names?
Like `test_wrong_absorber_dim` , etc.

**Review by DannyHindson — DISMISSED** [review]
This looks good to me. Thanks for making the changes to the tests.
Out of interest why is this validateInputs test separated out into its own unit? I've not seen two test suites for a single algorithm anywhere else. Maybe you are concerned a full validateInputs test would get large...

**Comment by gui-co** [comment]
Thanks for the review.
I think the plan was to separate the test of the processing from the test of the algorithm class itself. Because they are both small we could have merged them but this algorithm is still in development. We will see how they evolve. I have another PR pending concerning this algorithm, I can now add some checks in the test suite we just created here.

**Comment by gvardany** [comment]
I don't have a strong view. It is true that there is no other algorithm that has more than one unit test suite. So maybe indeed it's better to merge.

**Comment by gvardany** [comment]
Danny, we merged the tests, do you mind shipping again?

**Review by DannyHindson — APPROVED** [review]
Sorry missed that last change. Thanks for updating. All looks good now

**Pull request merged**
